### PR TITLE
Rewrite and simplify every indexer function

### PR DESCRIPTION
**Pull request by Kerollmops**
This is one the first steps toward making the engine easier to understand!

### CODE REVIEW

File: meilisearch-core/src/update/documents_addition.rs [hunk 1]
```diff
@@ -253,22 +267,31 @@ pub fn reindex_all_documents(writer: &mut heed::RwTxn<MainT>, index: &store::Ind
...
+                        *indexed_pos,
+                        number_of_words as u16,
+                    )?;
+                }
+            }
+
+            if schema.is_ranked(field_id) {
+                let number = value_to_number(&value).unwrap_or_default();
+                ranked_map.insert(document_id, field_id, number);
+            }
```
Comment (qdequele) [inline meilisearch-core/src/update/documents_addition.rs:294]: Proposal: Create an inline method for that. Its writer 2 times in the same file.

File: meilisearch-core/src/update/helpers.rs [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+            Value::Object(object) => {
+                for (key, value) in object {
+                    string.push_str(key);
+                    let _ = string.write_str(". ");
+                    internal_value_to_string(string, value);
+                    let _ = string.write_str(". ");
+                }
+            },
+        }
+    }
```
Comment (qdequele) [inline meilisearch-core/src/update/helpers.rs:70]: Is it conventional? Can't it be replaced by a Display implementation?

File: meilisearch-http/src/routes/document.rs [hunk 3]
```diff
@@ -6,6 +6,9 @@ use indexmap::IndexMap;
 use serde::Deserialize;
 use serde_json::Value;
 
+use meilisearch_core::{Error, serde::SerializerError};
+use meilisearch_core::update;
```
Comment (qdequele) [inline meilisearch-http/src/routes/document.rs:10]: They could be merged.

File: meilisearch-http/src/routes/document.rs [hunk 4]
```diff
@@ -42,7 +45,10 @@ async fn get_document(
         .open_index(&path.index_uid)
         .ok_or(ResponseError::index_not_found(&path.index_uid))?;
 
-    let document_id = meilisearch_core::serde::compute_document_id(&path.document_id);
+    let document_id = update::compute_document_id(&path.document_id);
+    if !update::validate_document_id(&path.document_id) {
+        return Err(Error::Serializer(SerializerError::InvalidDocumentIdFormat).into())
+    }
```
Comment (qdequele) [inline meilisearch-http/src/routes/document.rs:51]: Why `compute_document_id` did not return a Result? Instead of validating the document_id afterward.

Comment (qdequele) [reply meilisearch-http/src/routes/document.rs:51]: You do it 3 times in that file.

**Comment by qdequele** [comment]
Did you plan to create a Tracking issue about the MeiliSearch improvement?

File: meilisearch-core/src/update/helpers.rs (hunk 2 above)
Comment (Kerollmops) [reply meilisearch-core/src/update/helpers.rs:70]: No, Display must be implemented on a type, the type here is Value which is a foreign type, I can't impl two foreign interfaces together Also Value already impl Display in its way.

This function is really specific to the engine, it adds dots, is recursive and ignore certain values (i.e. nulls). I keep it like that, it is way easier to read than before and than a custom type to impl a custom Display on it.

File: meilisearch-core/src/lib.rs [hunk 5]
```diff
@@ -18,7 +18,7 @@ mod query_words_mapper;
 mod ranked_map;
 mod raw_document;
 mod reordered_attrs;
-mod update;
+pub mod update;
```
Comment (MarinPostma) [inline meilisearch-core/src/lib.rs:21]: can you put this after `pub mod store`

File: meilisearch-core/src/update/documents_addition.rs [hunk 6]
```diff
@@ -2,15 +2,20 @@ use std::collections::HashMap;
...
+use serde_json::Value;
 
 use crate::database::{MainT, UpdateT};
 use crate::database::{UpdateEvent, UpdateEventsEmitter};
 use crate::facets;
 use crate::raw_indexer::RawIndexer;
-use crate::serde::{extract_document_id, serialize_value_with_id, Deserializer, Serializer};
+use crate::serde::Deserializer;
+use crate::store::{DocumentsFields, DocumentsFieldsCounts};
 use crate::store;
```
Comment (MarinPostma) [inline meilisearch-core/src/update/documents_addition.rs:17]:
```suggestion
use crate::serde::Deserializer;
use crate::store::{self, DocumentsFields, DocumentsFieldsCounts};
```

File: meilisearch-core/src/update/documents_addition.rs [hunk 7]
```diff
@@ -104,10 +109,45 @@ pub fn push_documents_addition<D: serde::Serialize>(
...
+    documents_fields.put_document_field(writer, document_id, field_id, &serialized)?;
+
+    if let Some(indexed_pos) = schema.is_indexed(field_id) {
+        let number_of_words = index_value(indexer, document_id, *indexed_pos, value);
+        if let Some(number_of_words) = number_of_words {
+            documents_fields_counts.put_document_field_count(
+                writer,
+                document_id,
+                *indexed_pos,
+                number_of_words as u16,
```
[the remark this answers is not in the document]
Comment (Kerollmops) [reply meilisearch-core/src/update/documents_addition.rs:134]: I put the number of words found inside the field of a document, I must find a better name indeed.

File: meilisearch-core/src/update/documents_addition.rs [hunk 8]
```diff
@@ -253,21 +295,22 @@ pub fn reindex_all_documents(writer: &mut heed::RwTxn<MainT>, index: &store::Ind
...
             let (field_id, bytes) = result?;
-            let value: serde_json::Value = serde_json::from_slice(bytes)?;
+            let value: Value = serde_json::from_slice(bytes)?;
             ram_store.insert((document_id, field_id), value);
         }
 
-        for ((docid, field_id), value) in ram_store.drain() {
-            serialize_value_with_id(
+        // For each key-value pair in the document.
+        for ((document_id, field_id), value) in ram_store.drain() {
```
[the remark this answers is not in the document]
Comment (Kerollmops) [reply meilisearch-core/src/update/documents_addition.rs:303]: RamStore doesn't contain documents but postings lists, I can rename it `RamPostingsLists` or something else, but this is not the right PR to do so, this PR doesn't touch to this type.

File: meilisearch-core/src/update/helpers.rs [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+}
+
+/// Transforms the JSON Value type into a Number.
+pub fn value_to_number(value: &Value) -> Option<Number> {
+    use std::str::FromStr;
+
+    match value {
+        Value::Null => None,
+        Value::Bool(boolean) => Some(Number::Unsigned(*boolean as u64)),
+        Value::Number(number) => Number::from_str(&number.to_string()).ok(), // TODO improve that
```
[the remark this answers is not in the document]
Comment (Kerollmops) [reply meilisearch-core/src/update/helpers.rs:84]: Those are not the same Number types, this doens't compile but I'll improve that.

File: meilisearch-core/src/lib.rs (hunk 5 above)
Comment (Kerollmops) [reply meilisearch-core/src/lib.rs:21]: My mistake.

**Comment by Kerollmops** [comment]
@qdequele Do you think I can merge? I have another PR that comes just after.

**Pull request merged**